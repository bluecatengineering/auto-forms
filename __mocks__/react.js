module.exports = {
    ...jest.requireActual('react'),
    useCallback: jest.fn((fn) => fn),
    useContext: jest.fn(),
    useReducer: jest.fn((reducer, initial) => [initial, jest.fn()]),
};
