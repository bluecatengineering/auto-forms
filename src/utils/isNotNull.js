const isNotNull = value => (Array.isArray(value) ? value.some(isNotNull) : value !== null);

export default isNotNull;
