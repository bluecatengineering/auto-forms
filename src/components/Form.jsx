import React, {useCallback, useReducer} from 'react';
import PropTypes from 'prop-types';
import identity from 'lodash-es/identity';

import isNotNull from '../utils/isNotNull';

import FormContext from './FormContext';

const reducer = (state, {type, name, payload}) => {
	switch (type) {
		case 'SET_VALUE':
			return {...state, values: {...state.values, [name]: payload}};
		case 'SET_ERROR':
			return {...state, errors: {...state.errors, [name]: payload}};
		case 'SET_EXTRA':
			return {...state, extras: {...state.extras, [name]: payload}};
	}
	return state;
};

const getInitialState = ({initialValues, initialExtras, rules}) => ({
	initialValues,
	rules,
	values: initialValues,
	errors: {},
	extras: initialExtras,
});

const submitForm = ({rules, values, extras, initialValues}, dispatch, getActiveFields, extraValidation, onSubmit) => {
	const setError = (name, payload) => dispatch({type: 'SET_ERROR', name, payload});
	const keys = getActiveFields ? getActiveFields(values, extras) : Object.keys(values);
	const errors = keys.reduce((o, key) => {
		const rule = rules[key];
		if (rule) {
			o[key] = rule(values[key]);
		}
		return o;
	}, {});

	return Promise.resolve(extraValidation(errors, values, extras, initialValues)).then(errors => {
		const passed = Object.entries(errors).reduce(
			(r, [name, payload]) => (isNotNull(payload) ? (setError(name, payload), false) : r),
			true
		);

		if (passed) {
			const setErrors = errors => Object.entries(errors).forEach(([name, payload]) => setError(name, payload));
			onSubmit(values, {initialValues, extras, setErrors});
		}
	});
};

const Form = ({
	initialValues,
	initialExtras,
	rules,
	children,
	getActiveFields,
	extraValidation,
	onSubmit,
	...props
}) => {
	const [state, dispatch] = useReducer(reducer, {initialValues, initialExtras, rules}, getInitialState);
	const handleSubmit = useCallback(
		event => (event.preventDefault(), submitForm(state, dispatch, getActiveFields, extraValidation, onSubmit)),
		[state]
	);
	return (
		<FormContext.Provider value={{state, dispatch}}>
			<form {...props} onSubmit={handleSubmit}>
				{children}
			</form>
		</FormContext.Provider>
	);
};

Form.propTypes = {
	initialValues: PropTypes.object.isRequired,
	initialExtras: PropTypes.object,
	rules: PropTypes.object.isRequired,
	children: PropTypes.oneOfType([PropTypes.element, PropTypes.array]),
	getActiveFields: PropTypes.func,
	extraValidation: PropTypes.func,
	onSubmit: PropTypes.func,
};

Form.defaultProps = {
	initialExtras: {},
	extraValidation: identity,
};

export default Form;
