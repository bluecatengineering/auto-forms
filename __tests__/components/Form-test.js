import {useContext, useReducer, useEffect} from 'react';
import {shallow} from 'enzyme';

import Form from '../../src/components/Form';
import isNotNull from '../../src/utils/isNotNull';
import FormContext from '../../src/components/FormContext';

jest.unmock('../../src/components/Form');

const Test = () => {
	const {state, dispatch} = useContext(FormContext);
	useEffect(() => {
		if (state.values.test === 'foo') {
			dispatch({type: 'SET_VALUE', name: 'test', payload: 'new-value'});
			dispatch({type: 'SET_EXTRA', name: 'test', payload: 'new-extra'});
			dispatch({type: 'OTHER', name: 'test', payload: 'new-other'});
		}
	});
	return <div />;
};

describe('Form', () => {
	describe('rendering', () => {
		it('renders expected elements', () => {
			const wrapper = shallow(
				<Form id="test" initialValues={{test: 'foo'}} rules={{}}>
					<div />
				</Form>
			);
			expect(wrapper.getElement()).toMatchSnapshot();
		});
	});

	describe('behaviour', () => {
		it('reducer sets state depending on action', () => {
			const state = {foo: 'test'};
			shallow(<Form id="test" initialValues={{}} rules={{}}>
					<div />
				</Form>)
			const reducer = useReducer.mock.calls[0][0];
			expect(reducer(state, {type: 'SET_VALUE', name: 'other', payload: 'bar'}))
				.toEqual({foo: 'test', values: {other: 'bar'}});
			expect(reducer(state, {type: 'SET_ERROR', name: 'other', payload: 'bar-error'}))
				.toEqual({foo: 'test', errors: {other: 'bar-error'}});
			expect(reducer(state, {type: 'SET_EXTRA', name: 'other', payload: 'bar-extra'}))
				.toEqual({foo: 'test', extras: {other: 'bar-extra'}});
			expect(reducer(state, {type: 'RESET', payload: {other: 'bar'}}))
				.toEqual({other: 'bar'});
			expect(reducer(state, {type: 'FAKE', name: 'fake', payload: 'fake'}))
				.toEqual(state);
		});

		it ('calls onSubmit if validation passes', () => {
			const initialValues = {test: 'foo'};
			const onSubmit = jest.fn();
			const preventDefault = jest.fn();
			const event = {preventDefault};
			const dispatch = jest.fn();
			useReducer.mockReturnValue([{errors: {}, extras: {}, initialValues, values: initialValues}, dispatch])

			const wrapper = shallow(
				<Form id="test" initialValues={initialValues} rules={{}} onSubmit={onSubmit}>
					<div />
				</Form>
			);
			return wrapper.find('#test').prop('onSubmit')(event).then(() => {
				expect(preventDefault.mock.calls).toEqual([[]]);
				expect(onSubmit.mock.calls).toEqual([[initialValues, {
					initialValues,
					extras: {},
					setErrors: expect.any(Function)
				}]]);

				const setErrors = onSubmit.mock.calls[0][1].setErrors;
				setErrors({foo: "foo-test"});
				expect(dispatch.mock.calls).toEqual([[{name: "foo", payload: "foo-test", type: "SET_ERROR"}]])
			});
		})

		it ('does not call onSubmit if the validation fails', () => {
			const onSubmit = jest.fn();
			const initialValues = {test: 'foo', other: 'bar'};
			const rule = jest.fn().mockReturnValue('rule-error');
			const getActiveFields = jest.fn().mockReturnValue(['test']);
			const extraValidation = jest.fn().mockReturnValue({foo: 'foo-error', bar: 'bar-error'});
			isNotNull.mockReturnValueOnce(true).mockReturnValueOnce(false);
			const preventDefault = jest.fn();
			const event = {preventDefault};
			const dispatch = jest.fn();
			const state = {errors: {}, extras: {}, initialValues, values: initialValues}
			useReducer.mockReturnValue([state, dispatch])
			const wrapper = shallow(
				<Form
					id="test"
					initialValues={initialValues}
					rules={{test: rule}}
					getActiveFields={getActiveFields}
					extraValidation={extraValidation}
					onSubmit={onSubmit}
				/>
			);
			return wrapper.find('#test').prop('onSubmit')(event).then(() => {
				expect(rule.mock.calls).toEqual([['foo']]);
				expect(getActiveFields.mock.calls).toEqual([[initialValues, {}]]);
				expect(extraValidation.mock.calls).toEqual([[{test: 'rule-error'}, initialValues, {}, initialValues]]);
				expect(isNotNull.mock.calls).toEqual([['foo-error'], ['bar-error']]);
				expect(onSubmit.mock.calls).toEqual([])
			});
		})

		it ('resets the form state', () => {
			const initialValues = {test: 'foo'};
			const preventDefault = jest.fn();
			const event = {preventDefault};
			const dispatch = jest.fn();
			useReducer.mockReturnValue([{test: 'bar'}, dispatch])

			const wrapper = shallow(
				<Form id="test" initialValues={initialValues} rules={{}}>
					<div />
				</Form>
			);
			wrapper.find('#test').prop('onReset')(event);
			expect(preventDefault.mock.calls).toEqual([[]]);
			expect(dispatch.mock.calls).toEqual([[{
				type: 'RESET',
				payload: {errors: {}, extras: {}, initialValues, values: initialValues}
			}]]);
		})
	});
});
