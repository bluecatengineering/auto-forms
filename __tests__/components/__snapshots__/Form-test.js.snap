// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Form rendering renders expected elements 1`] = `
<Context.Provider
  value={
    Object {
      "dispatch": [Function],
      "rules": Object {},
      "state": Object {
        "errors": Object {},
        "extras": Object {},
        "initialValues": Object {
          "test": "foo",
        },
        "values": Object {
          "test": "foo",
        },
      },
      "submit": [Function],
    }
  }
>
  <form
    id="test"
    onSubmit={[Function]}
  >
    <div />
  </form>
</Context.Provider>
`;
